psd-web/app/models/user.rb

```diff
@@ -181,9 +185,54 @@ def invitation_expired?
     invited_at <= INVITATION_EXPIRATION_DAYS.days.ago
   end
 
-private
+  def two_factor_authentication_code_expired?
+    return false if !direct_otp_sent_at
+
+    (direct_otp_sent_at + User.direct_otp_valid_for) < Time.current
+  end
+
+  def two_factor_locked?
+    second_factor_attempts_locked_at && !two_factor_lock_expired?
+  end
+
+  def fail_two_factor_authentication!
+    return if max_login_attempts?
+
+    self.second_factor_attempts_count += 1
+    self.save
```

Comment:
Could use the Rails [`increment!` method](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-increment) here:

```suggestion
    self.increment!(:second_factor_attempts_count, 1)
```

The main difference is that validations and any callbacks will be skipped.

Reply:
Nice!